File: java/shared/java/org/signal/libsignal/zkgroup/backups/BackupLevel.java
//
// Copyright 2024 Signal Messenger, LLC.
// SPDX-License-Identifier: AGPL-3.0-only
//

package org.signal.libsignal.zkgroup.backups;

import java.util.Arrays;
import java.util.Map;
import java.util.function.Function;
import java.util.stream.Collectors;

public enum BackupLevel {
  // This must match the Rust version of the enum.
  MESSAGES(200),
  MEDIA(201);

  private static final Map<Integer, BackupLevel> LOOKUP =
      Arrays.stream(BackupLevel.values())
          .collect(Collectors.toMap(BackupLevel::getValue, Function.identity()));

  private final int value;

  BackupLevel(int value) {
    this.value = value;
  }

  int getValue() {
    return this.value;
  }

  public static BackupLevel fromValue(int value) {
    BackupLevel backupLevel = LOOKUP.get(value);
    if (backupLevel == null) {
      throw new IllegalArgumentException("Invalid backup level: " + value);
    }
    return backupLevel;
  }
}

File: java/shared/java/org/signal/libsignal/zkgroup/backups/BackupLevel.kt
//
// Copyright 2024 Signal Messenger, LLC.
// SPDX-License-Identifier: AGPL-3.0-only
//

package org.signal.libsignal.zkgroup.backups

public enum class BackupLevel {
  // This must match the Rust version of the enum.
  FREE(200),
  PAID(201),
  ;

  public val value: Int

  private constructor(value: Int) {
    this.value = value
  }

  public companion object {
    @JvmStatic
    public fun fromValue(value: Int): BackupLevel {
      // A linear scan is simpler than a hash lookup for a set of values this small.
      val found = BackupLevel.values().firstOrNull { it.value == value }
      return requireNotNull(found) { "Invalid backup level: $value" }
    }
  }
}
